students/elmar_m/lesson06/mailroom.py

```diff
@@ -0,0 +1,213 @@
+    '''
+    while True: 
+        dname = input('>> Please give donor name: ')
+        if not dname.isalpha():
+            print('>> only alphabetical characters in donor name allowed, please try again')
+            continue
+        else:
+            break
+    if dname in donors:
+        print('>>', dname, 'already in list')
+        return add_amount(dname)
+        # add_amount(dname)
+        # return True
+        
+    else:
+        print('>>', dname, 'not in list, adding it')
+        return add_amount(dname)
+        # add_amount(dname)
+        # return True
+    # return True
+
+
+def add_amount(donor):
+    '''
+    Add donation amount
+        
+        ARGS:
+    donor: name of donor
+    '''
+    while True:
+        amount = input('>> please add current donation (int or float, use \'.\' as decimal separator):\n>> ')
+        # make sure given donation values are positive: 
+        if '-' in amount:
+            print('>> only positive integers or floats allowed.')
+            continue
+        # Throw exception and ask again, if given value is not int or float:
+        try:
+            donors[donor].append(float(amount))
+            print(amount, 'added to donation list of', donor, ', thank you')     
+            return True
+            # break
+        except ValueError:
+            print('>> only numerical values allowed, please try again')
+
+
+def report():
+    '''
+    Print a formatted list of current donors and donations
+    '''
+    # get the highest number of digits to create formatstring accordingly, using a comprehension:
+    sumlist = [sum(i) for i in donors.values()]
+
+    maxn = len(str(max(sumlist)))
+    print('maxn ist:', maxn)
+
+    maxn += 2
+    fstring = '{:<20} ' + '|' + '{:>' + str(maxn) + '} ' + '|' + '{:>9}' + '|' + '{:>20}' 
+
+    print(fstring.format('Donor Name', 'Total', 'Num Gifts', 'Average Gift'))  
+    print('-' * (maxn + 54)) 
+
+    for i in donors:
+        a = fstring.format(i, sum(donors[i]), len(donors[i]),  sum(donors[i])//len(donors[i]) ) 
+        # print(fstring.format(i, sum(donors[i]), len(donors[i]),  sum(donors[i])//len(donors[i]) )) 
+        if a:
+            print(a)
+        else:
+            break
+    return True
+
+def efunc():
+    '''
+    Exit menu
+    '''
+    print('exiting.\n')
+    return 'exiting'
+
+
+main_d = {
+    # Dispatcher dictionary main menu
+    't' : thankyou,
+    'r' : report,
+    'q' : efunc,
+    }
+
+
+sub_d= {
+    #Dispatcher dictionary submenu
+    'l' : list_donors,
+    'a' : add,
+    'x' : efunc,
+    'm' : mail,
+    }
+
+
+def menu(p, d):
+    '''
+    Display menu to user. 
+    This function uses a "dispatcher dictionary".
+
+        ARGS:
+    p:  prompt which is shown to the user
+    d:  the dispatcher dictionary which holds the menu options
+    '''
+    try:
+        while True:
+            response = input(p)
+            if d[response]() == 'exiting':
+                break
+    except KeyError:
```

Comment:
your `KeyError` should be around line 204 and no need to call `menu(p, d)` as you are in a while True loop